src/capture.cpp

```diff
@@ -25,111 +29,42 @@ acquisition::Capture::~Capture(){
-        else{
-            ROS_FATAL_STREAM("  USB memory on system too low ("<<mem<<" MB)! Must be at least 1000 MB. Run: \nsudo sh -c \"echo 1000 > /sys/module/usbcore/parameters/usbfs_memory_mb\"\n Terminating...");
-            ros::shutdown();
-        }
-    } else {
-        ROS_FATAL_STREAM("Could not check USB memory on system! Terminating...");
-        ros::shutdown();
-    }
-
-    // default values for the parameters are set here. Should be removed eventually!!
-    exposure_time_ = 0 ; // default as 0 = auto exposure
-    soft_framerate_ = 20; //default soft framrate
-    ext_ = ".bmp";
-    SOFT_FRAME_RATE_CTRL_ = false;
-    LIVE_ = false;
-    TIME_BENCHMARK_ = false;
-    MASTER_TIMESTAMP_FOR_ALL_ = true;    
-    EXTERNAL_TRIGGER_ = false;
-    EXPORT_TO_ROS_ = false;
-    PUBLISH_CAM_INFO_ = false;
-    SAVE_ = false;
-    SAVE_BIN_ = false;
-    nframes_ = -1;
-    FIXED_NUM_FRAMES_ = false;
-    MAX_RATE_SAVE_ = false;
-    region_of_interest_set_ = false;
-    skip_num_ = 20; 
-    init_delay_ = 1; 
-    master_fps_ = 20.0;
-    binning_ = 1;
-    SPINNAKER_GET_NEXT_IMAGE_TIMEOUT_ = 2000;
-    todays_date_ = todays_date();
-    
-    dump_img_ = "dump" + ext_;
-
-    grab_time_ = 0;
-    save_time_ = 0;
-    toMat_time_ = 0;
-    save_mat_time_ = 0;
-    export_to_ROS_time_ = 0; 
-    achieved_time_ = 0;
-        
-    // decimation_ = 1;
-  
-    CAM_ = 0;
-
-    // default flag values
-
-    MANUAL_TRIGGER_ = false;
-    CAM_DIRS_CREATED_ = false;
-
-    GRID_CREATED_ = false;
-    VERIFY_BINNING_ = false;
-
-
-    //read_settings(config_file);
-    read_parameters();
-    
-    // Retrieve singleton reference to system object
-    ROS_INFO_STREAM("Creating system instance...");
-    system_ = System::GetInstance();
-
-    load_cameras();
- 
-    //initializing the ros publisher
-    acquisition_pub = nh_.advertise<spinnaker_sdk_camera_driver::SpinnakerImageNames>("camera", 1000);
-    //dynamic reconfigure
-    dynamicReCfgServer_ = new dynamic_reconfigure::Server<spinnaker_sdk_camera_driver::spinnaker_camConfig>(nh_pvt_);
-    
-    dynamic_reconfigure::Server<spinnaker_sdk_camera_driver::spinnaker_camConfig>::CallbackType dynamicReCfgServerCB_t;   
-
-    dynamicReCfgServerCB_t = boost::bind(&acquisition::Capture::dynamicReconfigureCallback,this, _1, _2);
-    dynamicReCfgServer_->setCallback(dynamicReCfgServerCB_t);
+void acquisition::Capture::onInit() {
+    NODELET_INFO("Initializing nodelet");
+    nh_ = this->getNodeHandle();
+    nh_pvt_ = this->getPrivateNodeHandle();
+    //it_.reset(new image_transport::ImageTransport(nh_));
+    it_ = std::shared_ptr<image_transport::ImageTransport>(new image_transport::ImageTransport(nh_));
+    // set values to global class variables and register pub, sub to ros
+    init_variables_register_to_ros();
+    init_array();
+    // calling capture::run() in a different thread
+    pubThread_.reset(new boost::thread(boost::bind(&acquisition::Capture::run, this)));
```

Comment:
Why call in different thread?

Reply:
onInit() is like constructor, we have to use fro minimal stuff and get out of it asap. the main code needs to be run in a different thread.